kotlin/android/app/src/main/java/dev/firezone/android/tunnel/DisconnectMonitor.kt

```diff
@@ -0,0 +1,37 @@
+/* Licensed under Apache 2.0 (C) 2024 Firezone, Inc. */
+import android.net.ConnectivityManager
+import android.net.LinkProperties
+import android.net.Network
+import dev.firezone.android.tunnel.TunnelService
+
+// None of the TunnelService lifecycle callbacks are called when a user disconnects the VPN
+// from the system settings. This class listens for network changes and shuts down the service
+// when the network is lost, which achieves the same effect.
+class DisconnectMonitor(private val tunnelService: TunnelService) : ConnectivityManager.NetworkCallback() {
+    private var vpnNetwork: Network? = null
+
+    // Android doesn't provide a good way to associate a network with a VPN service, so we
+    // have to use the IP addresses of the tunnel to determine if the network is our VPN.
+    override fun onLinkPropertiesChanged(
+        network: Network,
+        linkProperties: LinkProperties,
+    ) {
+        val ipv4Found = linkProperties.linkAddresses.find { it.address.hostAddress == tunnelService.tunnelIpv4Address }
+        val ipv6Found = linkProperties.linkAddresses.find { it.address.hostAddress == tunnelService.tunnelIpv6Address }
```

Comment:
Do we need to guard against `tunnelIpv4Address` being null here to avoid a potential `null == null` comparison? These are null before we start-up right? Might be more of a defense-in-depth thing because there is nothing to shut down if we aren't started up.

Reply:
Hm, I don't _think_ that's a possibility, but I'll refactor this just in case.